### DISCUSSION TITLE
Seaweedfs cluster: adding new servers and removing old ones

### DISCUSSION THREAD

**Discussion by JouHouFin** [seaweedfs/seaweedfs#4824] [Q&A] [answered]
Hello!

First, thanks for this project! 

We use seaweedfs in production:
version 30GB 3.51 4310e1fac4e59a4391159e6c349f45a8e1cd0601 linux amd64

Currently, our setup is as follows:
- four nodes, of which:
- three are masters
- all four are filers using embedded (ldb2) store
- all four are volume servers
- replication mode is 001

Three of those servers are old ones that need to be replaced with newer ones. We recently tested this process with our qa environment:
1. Added two new nodes to the cluster
2. Evacuated three old nodes (_lock; volumeServer.evacuate -node <volume-node-ip>:8080 --force; volumeServer.leave -node <volume-node-ip>:8080; unlock ; ## systemctl stop old weed-volumes_)
3. Shutdown three old nodes

While the volumes and the data were moved to the new nodes, metadata apparently did not and we ended up with missing attachments in our app. What we did after that:
1. Brought the old nodes back up and the attachments started showing up again
2. On on of the old nodes, fs.meta.save
3. On one of the new nodes, fs.meta.load
4. Shutdown OLD nodes again
5. Attachments working again with only NEW nodes up

So, we have a couple of questions:
1. What is the correct process of adding new servers and removing old ones?
2. Should this problem be mitigated by moving to an external filer store, for example postgresql?
3. How does swfs metadata work? Does it contain some kind of reference to the actual location of the file so that when master is contacted about a file, it points the caller the the right place?
4. What does fs.meta.save actually do? Is it enough to do it on one machine? Is it safe to do it on multiple machines and is it safe to do fs.meta.load on multiple machines with multiple saved metadata files (and is this needed)?
5. General tips about our setup? Does it sound sensible etc?

Right now, we cannot really replace the old servers until we can be sure that there will be no data loss.

**Comment by yeganemehr** [comment]
Hi
You did move your volumes correctly by running `volumeServer.evacuate` but you need to understand your filer metadata is in danger.
You using LevelDB which is local. Every filer of your 4 filers have a part of whole database. So each goes down some part of your files became inaccessible, if you lost one of your nodes, that part of your data is lost for ever no matter how you replicate your volumes.
You should consider [store replication](https://github.com/seaweedfs/seaweedfs/wiki/Filer-Store-Replication) and choose a [distributed store](https://github.com/seaweedfs/seaweedfs/wiki/Filer-Stores) .

**Comment by chrislusf** [comment, marked as answer]
* Each filer's local embedded filer store has the full copy of metadata. So your steps are correct. 
* It's better to move to an external filer store.
